mkosi/vmspawn.py

```diff
@@ -0,0 +1,123 @@
+
+
+def run_vmspawn(args: MkosiArgs, config: MkosiConfig) -> None:
+    if config.output_format not in (OutputFormat.disk, OutputFormat.directory):
+        die(f"{config.output_format} images cannot be booted in systemd-vmspawn")
+
+    if config.qemu_firmware == QemuFirmware.bios:
+        die("systemd-vmspawn cannot boot BIOS firmware images")
+
+    if config.qemu_cdrom:
+        die("systemd-vmspawn does not support CD-ROM images")
+
+    kernel = config.qemu_kernel
+
+    if kernel and not kernel.exists():
+        die(f"Kernel not found at {kernel}")
+
+    firmware = finalize_qemu_firmware(config, kernel)
+
+    if firmware == QemuFirmware.linux:
+        kernel = config.output_dir_or_cwd() / config.output_split_kernel
+        if not kernel.exists():
+            die(
+                f"Kernel or UKI not found at {kernel}",
+                hint="Please install a kernel in the image or provide a --qemu-kernel argument to mkosi vmspawn"
+            )
+
+    cmdline: list[PathString] = [
+        "systemd-vmspawn",
+        "--qemu-smp", config.qemu_smp,
+        "--qemu-mem", config.qemu_mem,
+        "--qemu-kvm", config.qemu_kvm.to_tristate(),
+        "--qemu-vsock", config.qemu_vsock.to_tristate(),
+        "--vtpm", config.qemu_swtpm.to_tristate(),
+        "--secure-boot", yes_no(config.secure_boot),
+    ]
+
+    if config.qemu_gui:
+        cmdline += ["--qemu-gui"]
+
+    cmdline += [f"--set-credential={k}:{v}" for k, v in config.credentials.items()]
+
+    with contextlib.ExitStack() as stack:
+        if config.ephemeral:
+            fname = stack.enter_context(copy_ephemeral(config, config.output_dir_or_cwd() / config.output))
+        else:
+            fname = config.output_dir_or_cwd() / config.output
+
+        if config.output_format == OutputFormat.disk and config.runtime_size:
+            apply_runtime_size(fname, config.runtime_size)
+
+        if (
+            kernel and
+            (KernelType.identify(kernel) != KernelType.uki or not config.architecture.supports_smbios(firmware))
+        ):
+            kcl = config.kernel_command_line + config.kernel_command_line_extra
+        else:
+            kcl = config.kernel_command_line_extra
+
+        for tree in config.runtime_trees:
+            target = Path("/root/src") / (tree.target or tree.source.name)
+            cmdline += ["--bind", f"{tree.source}:{target}"]
+
+        root = None
+        if kernel:
+            cmdline += ["--kernel", kernel]
+
+            if any(s.startswith("root=") for s in kcl):
+                pass
+            elif config.output_format == OutputFormat.disk:
+                # We can't rely on gpt-auto-generator when direct kernel booting so synthesize a root=
+                # kernel argument instead.
+                root = finalize_root(find_partitions(fname))
+                if not root:
+                    die("Cannot perform a direct kernel boot without a root or usr partition")
+
+                kcl += [root]
```

Comment:
This is another thing that vmspawn should do on its own eventually

Reply:
This definitely should be in vmspawn. I think I was a bit scared of using libfdisk before which is why I didn't implement it. I'm going to spend a couple hours fiddling with libfdisk because its a lot less scary than I thought.

Reply:
Got some advice from work and libblkid seems to be the way to go for this, I've found some sample code that works for a disk image generated by mkosi so I'm going to work on adding this to vmspawn now.

Reply:
Yeah that should work as well, we use both in systemd AFAIK.

Reply:
So one limitation of this is that afaict it won't be able to detect verity roothash's.
I think this is fine as vmspawn currently has no support for anything verity related?

Reply:
Once I've got this a bit polished I'll see how much effort it would be to add the libcryptsetup stuff required to find the verity roothash.

Reply:
Right this is now merged into my vmspawn PR and I have added it to mkosi as well.
ngl very happy with how mkosi vmspawn has turned out now that all this stuff has been pushed into vmspawn.
Changes in 8a0b8871.